cibyl/sources/source.py

```diff
@@ -74,35 +74,50 @@ def request_handler(*args, **kwargs):
 class Source(AttrDict):
     """Represents a data provider within a system."""
 
-    def __init__(self, name: str = None, driver: str = None, **kwargs):
+    def __init__(self: object,
+                 name: str = None,
+                 driver: str = None,
+                 **kwargs: Argument):
         kwargs.setdefault('enabled', True)
         kwargs.setdefault('_setup', False)
+        kwargs.setdefault('_down', False)
         kwargs.setdefault('priority', 0)
 
         super().__init__(name=name, driver=driver, **kwargs)
 
-    def is_setup(self):
+    def is_setup(self: object) -> bool:
         """Return wether the source has been setup."""
         return self._setup
 
-    def ensure_source_setup(self):
+    def ensure_source_setup(self: object) -> None:
         """Ensure that setup is called for the source. If setup was previously
         called, do nothing."""
         if not self.is_setup():
             self._setup = True
             self.setup()
 
+    def is_down(self: object) -> bool:
+        """Return wether the source has been teardown."""
+        return self._down
+
+    def ensure_teardown(self: object) -> None:
+        """Ensure that teardown is called for the source. If teardown was
+        previously called, do nothing."""
+        if not self.is_down():
+            self._down = True
+            self.teardown()
+
     @abstractmethod
-    def setup(self):
+    def setup(self: object) -> None:
```

Comment:
This is really redundant. The docstring already does not have a \:return: entry, that already tells you it is a void function.

Reply:
Although from a readability standpoint it's obvious that this function returns None, if we want to enforce type checks in CI, this annotations is quite standard, see for reference the example Class annotation in mypy docs: https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html#classes

Reply:
The docstring already documents the function, I do not see why we have to do it twice. 

I have my doubts about enforcing type checking. I would have no problem with it if at the beginning of the project we had decided on one style for the whole codebase, but at this point we are using different ways to achieve the same thing. I have been using docstrings to document my code because our really strict 80 column rule makes type hints occupy way too much room. The two IDEs that I have used: VSCode and PyCharm do not care about what you use, and provide the type hints for either style. So really, aside from consistency, there is not much to gain by having it as a rigid CI rule.

Reply:
That's a good point, I personally don't have any strong feeling in favour or against enforcing it on CI, but we should make sure we raise any drawbacks in our team discussions when we decide whether we go for it or not.

Reply:
As in the other comment: Let's discuss it in the meeting then because there's already a lot of methods with this definition.
Because even if it's redundant there are already a lot of parts in all the project where:

- Methods with docstrings.
- Methods without docstrings.
- Methods with type hints.
- Methods without type hints.
- Methods with docstrings and type hints.
- Methods without docstrings and without type hints.

I understand the redundancy but we need to decide what to do, because it's a little bit messy.

Reply:
Definitely let's agree before making all the changes, but I'd like to point out that type hints and docstrings, while they have some overlap, are not really equivalent. Docstrings are documentation (the types in the docstrings can be used by sphinx to generate a browsable package description) while the type hints enforced with mypy can be useful to detect bugs that in statically typed languages would be caught by the compiler.